Plutarch/FFI.hs

```diff
@@ -0,0 +1,145 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.FFI (
+  type (>~<),
+  foreignExport,
+  foreignImport,
+  unsafeForeignExport,
+  unsafeForeignImport,
+) where
+
+import Data.ByteString (ByteString)
+import Data.Kind (Constraint, Type)
+import Data.Text (Text)
+import GHC.Generics (C, D, K1, M1, Meta (MetaData), Rep, S)
+import GHC.TypeLits (TypeError)
+import qualified GHC.TypeLits as TypeLits
+import qualified Generics.SOP as SOP
+import Plutarch.Bool (PBool)
+import Plutarch.Builtin (PAsData, PData)
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  ClosedTerm,
+  PDelayed,
+  PType,
+  RawTerm (RCompiled),
+  Term (Term),
+  TermResult (TermResult),
+  asClosedRawTerm,
+  compile',
+  (:-->),
+ )
+import Plutarch.Internal.PlutusType (PlutusType (PInner))
+import Plutarch.String (PString)
+import Plutarch.Unit (PUnit)
+import Plutus.V1.Ledger.Scripts (Script (unScript), fromCompiledCode)
+import PlutusTx.Builtins.Internal (BuiltinBool, BuiltinByteString, BuiltinData, BuiltinUnit)
+import PlutusTx.Code (CompiledCode, CompiledCodeIn (DeserializedCode))
+import PlutusTx.Prelude (BuiltinString)
+import UntypedPlutusCore (fakeNameDeBruijn)
+import qualified UntypedPlutusCore as UPLC
+
+data ForallPhantom :: Type
+data PhorallPhantom :: PType
+
+data Delayed :: Type -> Type
+
+-- | Compile and export a Plutarch term so it can be used by `PlutusTx.applyCode`.
+foreignExport :: p >~< t => ClosedTerm p -> CompiledCode t
+foreignExport = unsafeForeignExport
+
+-- | Import compiled UPLC code (such as a spliced `PlutusTx.compile` result) as a Plutarch term.
+foreignImport :: p >~< t => CompiledCode t -> ClosedTerm p
+foreignImport = unsafeForeignImport
+
+-- | Seriously unsafe, may fail at run time or result in unexpected behaviour in your on-chain validator.
+unsafeForeignExport :: ClosedTerm p -> CompiledCode t
+unsafeForeignExport t = DeserializedCode program Nothing mempty
+  where
+    program =
+      UPLC.Program () (UPLC.Version () 1 0 0) $
+        UPLC.termMapNames fakeNameDeBruijn $
+          compile' $
+            asClosedRawTerm t
+
+-- | Seriously unsafe, may fail at run time or result in unexpected behaviour in your on-chain validator.
+unsafeForeignImport :: CompiledCode t -> ClosedTerm p
+unsafeForeignImport c = Term $ const $ TermResult (RCompiled $ UPLC.toTerm $ unScript $ fromCompiledCode c) []
+
+-- | Equality of inner types - Plutarch on the left and Haskell on the right.
+type p >~< t = PlutarchInner p PhorallPhantom ~~ PlutusTxInner t ForallPhantom
+
+type family a ~~ b :: Constraint where
+  ForallPhantom ~~ _ = ()
+  _ ~~ ForallPhantom = ()
+  Delayed a ~~ Delayed b = a ~~ b
+  a ~~ b = a ~ b
+
+type family PlutarchInner (p :: PType) (any :: PType) :: Type where
```

Comment:
I still don't understand why this type is so complex. What is wrong with something like
```haskell
type family PlutarchToPlutusTx (p :: PType) = t | t -> p where
  PlutarchToPlutusTx PBool = BuiltinBool
  PlutarchToPlutusTx PInteger = Integer
  PlutarchToPlutusTx PByteString = BuiltinByteString
  PlutarchToPlutusTx (a :--> b) = PlutarchToPlutusTx a -> PlutarchToPlutusTx b
```
etc.

Right now AFAICT you have two type families that convert to some intermediate types *that are never actually used* but only used for checking equality.

Reply:
In addition, what does `any` do?

Reply:
There's also
```
PlutarchToPlutusTx (PPair a b) = (PlutarchToPlutusTx a, PlutarchToPlutusTx b)
```
But `PPair a b >~< (a, b)` is just one of many possible ADT pairings between Plutarch and Haskell. We can't know them all in advance, because some may be user-defined. We do know however that they will be Scott-encoded, so the two type families that map them to Scott encodings are sufficient to cover them all.

The `any` serves the same purpose as the second type parameter of `PInner`, it acts as a universal type variable. This solution only covers types with a single forall-bound type variable, but it seems sufficient for most uses.

Reply:
So what's the purpose of having the intermediate types?

Reply:
The intermediate types are used only for checking equality of the two compiled types. Say a user has existing Haskell code dealing with
```
data MyRedeemer = MyRedeemer PubKeyHash Value
```
They want to be able to convert between `MyRedeemer` on the Haskell/PlutusTx side and
```
data PMyRedeemer s = PMyRedeemer (Term s PPubKeyHash) (Term s PValue)
```
on the Plutarch side. We could export an *open* type family `PlutarchToPlutusTx` as you suggest so that the user can add a new equivalence between the two redeemer types, but we wouldn't be able to actually test its correctness. The only proof of correctness is the equivalence of their Scott encodings. Thus the two type families.